horovod/mxnet/__init__.py

```diff
@@ -167,17 +182,22 @@ def _allreduce_grads(self):
 
                 for entries in entries_by_dtype.values():
                     grads, names = zip(*entries)
-                    grouped_allreduce_(tensors=grads, average=False, name="{}:{}".format(names[0], names[-1]), priority=-i,
+                    tensors_compressed, ctxs = zip(*[self._compression.compress(grad) for grad in grads])
+                    grouped_allreduce_(tensors=tensors_compressed, average=False, name="{}:{}".format(names[0], names[-1]), priority=-i,
                                        prescale_factor=1.0 / self._gradient_predivide_factor)
+                    grads = [self._compression.decompress(t, ctx) for t, ctx in zip(tensors_compressed, ctxs)]
         else:
             # In MXNet 2.0, param.name is no longer unique.
             # Meanwhile, since horovod requires Python 3.6, there is no need to sort
             # self._params as enumerating a python dict is always deterministic.
             for i, param in enumerate(self._params):
                 if param.grad_req != 'null':
-                    allreduce_(param.list_grad()[0], average=False,
+                    tensor_compressed, ctx = self._compression.compress(param.list_grad()[0])
+                    allreduce_(tensor_compressed, average=False,
                                name=self._prefix + str(i), priority=-i,
                                prescale_factor=1.0 / self._gradient_predivide_factor)
+                    param._grad[0][:] = self._compression.decompress(tensor_compressed, ctx)
```

Comment:
This is a place where it looks like you are actually modifying the parameter gradients. However, this is using a copy which isn't ideal, especially since this code is called even if someone isn't using compression.